File: Engine/Source/Native/Core/Mono/ManagedModule.cpp
#include "ManagedModule.h"

namespace Vanguard
{
	ManagedModule::ManagedModule(ManagedObject* aManagedObject)
	{
		managedObject = aManagedObject;
	}

	void ManagedModule::LoadModule(){}
	void ManagedModule::UnloadModule(){}

	String ManagedModule::GetModuleName()
	{
		MonoObject* returnValue = managedObject->CallMethod("GetModuleName");
		const char* result = (const char*)mono_object_unbox(returnValue);
		return result;
	}

	String ManagedModule::GetModuleType()
	{
		MonoObject* returnValue = managedObject->CallMethod("GetModuleType");
		const char* result = (const char*)mono_object_unbox(returnValue);
		return result;
	}
}

File: Engine/Source/Managed/ManagedModule/ManagedModule.cs
using System;

namespace Vanguard
{
    public class ManagedModule : IModule
    {
        public void Test()
        {
           //Log.Message("Static Calll", "Modules");
        }

        public void LoadModule()
        {
            Log.Message("Loaded Managed Module: " + GetModuleName(), "Modules");
        }

        public void UnloadModule()
        {
            Log.Message("Unloaded Managed Module: " + GetModuleName(), "Modules");
        }

        public String GetModuleName()
        {
            return "ManagedModule";
        }

        public String GetModuleType()
        {
            return "Managed";
        }
    }
}
